File: AlgorithmsOnGraphs-UCSD/week2/strongly_connected.py
#Uses python3

import sys

sys.setrecursionlimit(200000)

def dfs(adj, used, order, x):
    #write your code here
    used[x] = True
    for node in adj[x]:
        if not used[node]:
            dfs(adj, used, order, node)
    order.append(x)

def transpose_graph(adj):
    redges = [[] for _ in range(len(adj))]
    for i in range(len(adj)):
        for j in range(len(adj[i])):
            redges[adj[i][j]].append(i)
    return redges

'''    
def toposort(adj):
    used = [0] * len(adj)
    order = []
    #write your code here
    for i in range(len(adj)):
        if not used[i]:
            dfs(adj, used, order, i)
    return order[::-1]
'''

def number_of_strongly_connected_components(adj):
    result = 0
    #write your code here
    traversed = len(adj) * [False]
    stack = []
    for i in range(len(adj)):
        if not traversed[i]:
            dfs(adj, traversed, stack, i) #get ordering via stack and toposort
    #reverse graph
    radj = transpose_graph(adj)
    traversed = len(adj) * [False]
    while stack:
        curr = stack.pop()
        if not traversed[curr]:
            elestack = []
            dfs(radj, traversed, elestack, curr)
            result += 1
    return result

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    print(number_of_strongly_connected_components(adj))

File: AlgorithmsOnGraphs-UCSD/week2/strongly_connected.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <stack>

using std::vector;
using std::pair;
using std::stack;

void topological_sort(vector<vector<int> > &adj, vector<bool> &used, stack<int> &s, int x) {
    //write your code here
    used[x] = true;
    for(auto it = adj[x].begin(); it != adj[x].end(); it++){
        if(!used[*it]){
            used[*it] = true;
            topological_sort(adj, used, s, *it);
        }
    }
    // All vertices reachable from x are processed by now, push x to Stack
    s.push(x);
}
//reverse edges
vector<vector<int> > transpose_graph(vector<vector<int> > adj){
    vector<vector<int> > rEdges(adj.size(), vector<int>());

    for(int i = 0; i < adj.size(); i++){
        for(int j = 0; j < adj[i].size(); j++){
            rEdges[adj[i][j]].push_back(i);
        }
    }
    return rEdges;
}

int number_of_strongly_connected_components(vector<vector<int> > adj) {
    int result = 0;
    //write your code here
    stack<int> topo;
    vector<bool> visited(adj.size(), false);
    for(int i = 0; i < adj.size(); i++){
        if(!visited[i])
            topological_sort(adj, visited, topo, i);
    }

    vector<vector<int> > rGraph = transpose_graph(adj);
    visited.clear();
    visited.assign(rGraph.size(), false);
    visited.shrink_to_fit();
    
    while(!topo.empty()){
        int curr = topo.top();
        topo.pop();
        if(!visited[curr]){
            stack<int> elementStack;
            topological_sort(rGraph, visited, elementStack, curr);
            result++;
        }
    }
    return result;
}

int main() {
    size_t n, m;
    std::cin >> n >> m;
    vector<vector<int> > adj(n, vector<int>());
    for (size_t i = 0; i < m; i++) {
        int x, y;
        std::cin >> x >> y;
        adj[x - 1].push_back(y - 1);
    }
    std::cout << number_of_strongly_connected_components(adj);
}
